### PR TITLE
Migrate apex.parallel.SyncBatchNorm channels_last to pytorch

### PR DESCRIPTION
**Pull request by xwang233**
per title

This PR did
- Migrate `apex.parallel.SyncBatchNorm` channels_last to pytorch `torch.nn.SyncBatchNorm`
- Fix a TODO here by fusing `sum`, `div` kernels into backward elementwise kernel
https://github.com/pytorch/pytorch/blob/b167402e2e66a663cd9913885552929b4c045ffa/torch/nn/modules/_functions.py#L76-L95

Todo
- [x] Discuss a regression introduced in https://github.com/pytorch/pytorch/pull/37133#discussion_r512530389, which is the synchronized copy here 
https://github.com/pytorch/pytorch/blob/b167402e2e66a663cd9913885552929b4c045ffa/torch/nn/modules/_functions.py#L32-L34

**Comment**: This PR uses apex version for the size check. Test passed and I haven't seen anything wrong so far.

- [x] The restriction to use channels_last kernel will be like this
```
inline bool batch_norm_use_channels_last_kernels(const at::Tensor& self) {
  return self.is_contiguous(at::MemoryFormat::ChannelsLast) || self.ndimension() == 2;
}
```
I think we can relax that for channels_last_3d as well?

**Comment**: we don't have benchmark for this now, will check this and add functionality later when needed.
- [x] Add test
- [x] Add benchmark 

Detailed benchmark is at https://github.com/xwang233/code-snippet/tree/master/syncbn-channels-last

Close #50781

### CODE REVIEW

**Comment by xwang233** [comment]
cc @ptrblck @jjsjann123 @ngimel

**Comment by ngimel** [comment]
Cool, can you post benchmarks comparing to apex?

**Comment by pritamdamania87** [comment]
cc @lly-zero-one

**Comment by xwang233** [comment]
The detailed benchmark and raw data is at https://github.com/xwang233/code-snippet/tree/master/syncbn-channels-last. 

For 2D and 4D tensors on V100 x8 (relative perf is similar on A100 x8), the **kernel execution time** (not including NCCL reduction/gather, kernel launch overhead, or tensor memory format transformation):

new channels_last vs master contiguous 


new channels_last vs apex channels_last

**Comment by lly-zero-one** [comment]
> @lly-zero-one Would it be possible to test this PR with some of our ClassyVision workflows to see the potential benefit?

@pritamdamania87  In ClassyVision flow, it is using the Apex. Maybe we should ask CV team to change their flow.

**Comment by xwang233** [comment]
@ngimel The CLA is ready.

**Comment by VitalyFedyunin** [comment]
Hi! Can you please rebase, thanks

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 1]
```diff
@@ -857,4 +957,690 @@ std::tuple<Tensor, Tensor> batch_norm_update_stats_cuda_template(
...
+std::tuple<Tensor, Tensor> batch_norm_stats_channels_last_cuda_template(const Tensor& input, double epsilon) {
+  using accscalar_t = at::acc_type<scalar_t, true>;
+
+  const auto stride = input.size(1);
+  const auto reduction_size = input.numel() / stride;
+
+  auto scalar_type = input.scalar_type() == at::kHalf ? at::kFloat : input.scalar_type();
+  auto option = input.options().dtype(scalar_type);
+
+  at::Tensor out_invstd = at::empty({stride}, option);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Normalization.cuh:1404]: you can call at::native::empty_cuda here for slightly better perf, similar in other places

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 2]
```diff
@@ -857,4 +957,690 @@ std::tuple<Tensor, Tensor> batch_norm_update_stats_cuda_template(
...
+      <<<grid, block, 0, at::cuda::getCurrentCUDAStream()>>>(
+      input.data_ptr<scalar_t>(),
+      out_mean.data_ptr<accscalar_t>(),
+      out_invstd.data_ptr<accscalar_t>(),
+      staging_data_ptr,
+      semaphores_ptr,
+      reduction_size,
+      stride,
+      epsilon);
+  AT_CUDA_CHECK(cudaGetLastError());
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Normalization.cuh:1430]: C10_CUDA_KERNEL_LAUNCH_CHECK, similar in other places

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 3]
```diff
@@ -857,4 +957,690 @@ std::tuple<Tensor, Tensor> batch_norm_update_stats_cuda_template(
...
+    at::Tensor& output,
+    const at::Tensor& input,
+    const at::Tensor& weight,
+    const at::Tensor& shift,  // bias of BN
+    const at::Tensor& mean,
+    const at::Tensor& inv_std,
+    double epsilon,
+    const at::optional<at::Tensor>& z,  // bias after BN
+    const bool fuse_relu) {
+  const auto stride = input.size(1);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Normalization.cuh:1445]: you probably need to check that input is at least 2d. After than input.sizes()[1] is slightly faster

File: aten/src/ATen/native/cuda/Normalization.cu [hunk 4]
```diff
@@ -91,13 +101,19 @@ std::tuple<Tensor, Tensor> batch_norm_stats_cuda(const Tensor& self, double epsi
...
+  auto output = at::empty_like(self, self.suggest_memory_format());
   batch_norm_elemt_cuda_out(output, self, weight, bias, mean, invstd, epsilon);
   return output;
 }
 
 Tensor& batch_norm_elemt_cuda_out(Tensor& output, const Tensor& self, const Tensor& weight, const Tensor& bias,
                              const Tensor& mean, const Tensor& invstd, double epsilon) {
+  if (at::cuda::detail::canUse32BitIndexMath(self) && batch_norm_use_channels_last_kernels(self)){
+    batch_norm_elemt_channels_last_cuda_template(output, self, weight, bias, mean, invstd, epsilon,
+      /* (bias after BN) z = */ c10::nullopt, /* fuse_relu = */ false);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Normalization.cu:113]: if these options are never used, please remove them.

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 5]
```diff
@@ -857,4 +957,690 @@ std::tuple<Tensor, Tensor> batch_norm_update_stats_cuda_template(
...
+          input.data_ptr<scalar_t>(),
+          z.has_value() ? z.value().data_ptr<scalar_t>() : nullptr,
+          mean.data_ptr<accscalar_t>(),
+          inv_std.data_ptr<accscalar_t>(),
+          weight.defined() ? weight.data_ptr<accscalar_t>() : nullptr,
+          shift.defined() ? shift.data_ptr<accscalar_t>() : nullptr,
+          output.data_ptr<scalar_t>(),
+          reduction_size,
+          stride,
+          fuse_relu);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Normalization.cuh:1468]: hard code `false` here, and nullptr for `z`, no need to change the kernel

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 6]
```diff
@@ -857,4 +957,690 @@ std::tuple<Tensor, Tensor> batch_norm_update_stats_cuda_template(
...
+  const auto reduction_size = input.numel() / stride;
+
+  dim3 block;
+  dim3 grid;
+  flexible_launch_configs(reduction_size, stride, block, grid);
+
+  auto stream = at::cuda::getCurrentCUDAStream();
+
+  if (input.scalar_type() == at::kHalf && weight.defined() && weight.scalar_type() == at::kFloat) {
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.scalar_type(), "batchnorm_forward", [&] {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Normalization.cuh:1455]: please also enable bfloat16 dispatch

File: torch/nn/modules/_functions.py [hunk 7]
```diff
@@ -7,7 +7,12 @@ class SyncBatchNorm(Function):
 
     @staticmethod
     def forward(self, input, weight, bias, running_mean, running_var, eps, momentum, process_group, world_size):
-        input = input.contiguous()
+        if not input.is_contiguous(memory_format=torch.channels_last):
+            input = input.contiguous()
+
+        size = int(input.numel() / input.size(1))
```
Comment (ngimel) [inline torch/nn/modules/_functions.py:13]: is there a check anywhere that input.size(1) is not 0? `input.numel() // input.size(1)`

File: torch/nn/modules/_functions.py [hunk 8]
```diff
@@ -45,7 +46,7 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
             count_all.view(-1)
         )
 
-        self.save_for_backward(input, weight, mean, invstd, count_all)
+        self.save_for_backward(input, weight, mean, invstd, count_all.to(torch.int32))
```
Comment (ngimel) [inline torch/nn/modules/_functions.py:49]: I think there were recent changes to make `count` floating point to call a single allgather? Please check and update accordingly

File: torch/nn/modules/_functions.py [hunk 9]
```diff
@@ -7,7 +7,12 @@ class SyncBatchNorm(Function):
 
     @staticmethod
     def forward(self, input, weight, bias, running_mean, running_var, eps, momentum, process_group, world_size):
-        input = input.contiguous()
+        if not input.is_contiguous(memory_format=torch.channels_last):
+            input = input.contiguous()
```
Comment (ngimel) [inline torch/nn/modules/_functions.py:11]: I don't see anywhere where it's checked that weight is contiguous - is there such a place?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 10]
```diff
@@ -2453,11 +2453,13 @@ def _prepare_dummy_data(self, local_bs):
             return global_bs, input_cpu, target, loss
 
         # END TO END TEST FOR DISTRIBUTEDDATAPARALLEL
-        def _test_DDP_helper(self, model, input_var, target, loss, scale_factor=1.0):
+        def _test_DDP_helper(self, model, input_var, target, loss, scale_factor=1.0, check_memory_format=None):
```
Comment (ngimel) [inline torch/testing/_internal/distributed/distributed_test.py:2456]: `memory_format=None`, not `check_memory_format` - it's semantically unclear.

File: aten/src/ATen/native/cuda/Normalization.cuh (hunk 1 above)
Comment (xwang233) [reply aten/src/ATen/native/cuda/Normalization.cuh:1396]: Using `at::native::empty_cuda` seems to cause a few problems, so I switched back to `at::empty` plus `input.options()`.

File: aten/src/ATen/native/cuda/Normalization.cuh (hunk 3 above)
Comment (xwang233) [reply aten/src/ATen/native/cuda/Normalization.cuh]: It is checked in python https://github.com/pytorch/pytorch/blob/bf4fcab68104b333c81e39d142dccc7ae7a6f1d7/torch/nn/modules/batchnorm.py#L474-L477

File: torch/nn/modules/_functions.py (hunk 7 above)
Comment (xwang233) [reply torch/nn/modules/_functions.py]: I added a check in python here https://github.com/pytorch/pytorch/pull/46906/commits/d68394e0dd96d4db2f374e5787e279f13bb9fcc9

File: torch/nn/modules/_functions.py (hunk 8 above)
Comment (xwang233) [reply torch/nn/modules/_functions.py:50]: Hmm, I didn't see a relevant change. If you mean this https://github.com/pytorch/pytorch/pull/50126/files, then it should still be fine.

File: torch/nn/modules/_functions.py (hunk 9 above)
Comment (xwang233) [reply torch/nn/modules/_functions.py:11]: I think `weight` is initialized as a 1d tensor. Do we need to add a `weight = weight.contiguous()` here?

https://github.com/pytorch/pytorch/blob/bf4fcab68104b333c81e39d142dccc7ae7a6f1d7/torch/nn/modules/batchnorm.py#L40

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 11]
```diff
@@ -155,6 +155,100 @@ __device__ scalar_t reduce(Op op, PTA tensor, int plane) {
...
+__host__ __forceinline__ int h_last_pow2(unsigned int n) {
+    n |= (n >>  1);
+    n |= (n >>  2);
+    n |= (n >>  4);
+    n |= (n >>  8);
+    n |= (n >> 16);
+    return n - (n >> 1);
+}
+
+__host__ int div_ru(int x, int y) {
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/Normalization.cuh:163]: round up? I'll just rename the div.

File: aten/src/ATen/native/cuda/Normalization.cuh [hunk 12]
```diff
@@ -857,4 +957,690 @@ std::tuple<Tensor, Tensor> batch_norm_update_stats_cuda_template(
...
+  auto count_th = count[0];
+
+  // block-wise reduction with shared memory (since reduction cannot be done within a warp)
+  static __shared__ accscalar_t shmem_mean[MAX_BLOCK_SIZE];
+  static __shared__ accscalar_t shmem_m2n[MAX_BLOCK_SIZE];
+  static __shared__ int shmem_count[MAX_BLOCK_SIZE];
+
+  welford_merge_block_vertical(count_th, mean_th, m2_th, shmem_count, shmem_mean, shmem_m2n);
+
+  // grid reduction if needed (coop launch used at the first place)
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/Normalization.cuh:1043]: Wait, we don't really need coop launch since we are not busy spinning... I'll remove the comment.

File: torch/nn/modules/_functions.py (hunk 9 above)
Comment (ngimel) [reply torch/nn/modules/_functions.py:11]: Even if `weight` is initialized as contiguous, there's no guarantee it's not changed later.

Comment (albanD) [reply torch/nn/modules/_functions.py:11]: The call to contiguous should already be a no-op if the Tensor is already contiguous. Maybe you want to pass the memory_format arg directly to the contiguous call?

Comment (ngimel) [reply torch/nn/modules/_functions.py:11]: IIUC, the logic here is that if input is not channels-last contiguous, regular nchw path will be taken.

**Review by ngimel — APPROVED** [review]
lgtm, let's wait for the tests

**Comment by ngimel** [comment]
bc compat error is real

**Comment by xwang233** [comment]
> bc compat error is real

Yes, it is intentional. We fused mean calculations of `sum / num_channels` into the apply normalization kernel, so that reduces the number of kernels launched.

**Comment by ngimel** [comment]
I understand, but then you should add it to exceptions in bc compat test

**Comment by ngimel** [comment]
I don't want to delay this PR, but consider making functions like `batch_norm_backward_elemt` private by renaming to `_batch_norm_backward_elemt` in a follow up.

**Comment by osalpekar** [comment]
Hey @xwang233 @ngimel , it looks like the `pytorch_linux_backward_compatibility_check_test` test has been failing on master since this PR was merged. Have the appropriate exceptions been added to the bc compat test in this PR?

**Comment by xwang233** [comment]
Yes, bc compact exception was added here https://github.com/pytorch/pytorch/pull/46906/commits/c6c680a6c88de1762f31b9fbd9eb7f929b219587

The error message you saw was due to a revert of the previous commit in master branch before this one

```
Mar 04 00:53:49 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not. 
Mar 04 00:53:49 
Mar 04 00:53:49 Broken ops: [
Mar 04 00:53:49 	aten::_lstsq_helper(Tensor a, Tensor b, float cond, str? driver_name) -> (Tensor, Tensor, Tensor)
Mar 04 00:53:49 	aten::linalg_lstsq(Tensor self, Tensor b, float? cond=None, *, str? driver=None) -> (Tensor solution, Tensor residuals, Tensor rank, Tensor singular_values)
Mar 04 00:53:49 ]
```
